zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -20,344 +20,487 @@
-
+    getZkClient(path).watchForData(path);
   }
 
   @Override
   public List<String> watchForChilds(String path) {
-    return null;
+    return getZkClient(path).watchForChilds(path);
   }
 
   @Override
   public long getCreationTime(String path) {
-    return 0;
+    return getZkClient(path).getCreationTime(path);
   }
 
   @Override
   public List<OpResult> multi(Iterable<Op> ops) {
+    throwUnsupportedOperationException();
     return null;
   }
 
   @Override
   public boolean waitUntilConnected(long time, TimeUnit timeUnit) {
+    throwUnsupportedOperationException();
     return false;
   }
 
   @Override
   public String getServers() {
+    throwUnsupportedOperationException();
     return null;
   }
 
   @Override
   public long getSessionId() {
-    return 0;
+    // Session-aware is unsupported.
+    throwUnsupportedOperationException();
+    return 0L;
   }
 
   @Override
   public void close() {
+    if (isClosed()) {
+      return;
+    }
 
+    LOG.info("Closing {}.", FEDERATED_ZK_CLIENT);
+
+    synchronized (_zkRealmToZkClientMap) {
+      _zkRealmToZkClientMap.values().forEach(ZkClient::close);
+      _zkRealmToZkClientMap.clear();
+      _isClosed = true;
+    }
   }
 
   @Override
   public boolean isClosed() {
-    return false;
+    return _isClosed;
   }
 
   @Override
   public byte[] serialize(Object data, String path) {
-    return new byte[0];
+    return getZkClient(path).serialize(data, path);
   }
 
   @Override
   public <T> T deserialize(byte[] data, String path) {
-    return null;
+    return getZkClient(path).deserialize(data, path);
   }
 
   @Override
   public void setZkSerializer(ZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = new BasicZkSerializer(zkSerializer);
+    _zkRealmToZkClientMap.values()
+        .forEach(zkClient -> zkClient.setZkSerializer(_pathBasedZkSerializer));
   }
 
   @Override
   public void setZkSerializer(PathBasedZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = zkSerializer;
+    _zkRealmToZkClientMap.values().forEach(zkClient -> zkClient.setZkSerializer(zkSerializer));
   }
 
   @Override
   public PathBasedZkSerializer getZkSerializer() {
-    return null;
+    return _pathBasedZkSerializer;
+  }
+
+  private String create(final String path, final Object dataObject, final List<ACL> acl,
+      final CreateMode mode, final String expectedSessionId) {
+    if (mode.isEphemeral()) {
+      throwUnsupportedOperationException();
+    }
+
+    // Create mode is not session-aware, so the node does not have to be created
+    // by the expectedSessionId.
+    return getZkClient(path).create(path, dataObject, acl, mode);
+  }
+
+  private ZkClient getZkClient(String path) {
+    // If FederatedZkClient is closed, should not return ZkClient.
+    if (isClosed()) {
+      throw new IllegalStateException(FEDERATED_ZK_CLIENT + " is closed!");
+    }
+
+    String zkRealm = getZkRealm(path);
+    if (!_zkRealmToZkClientMap.containsKey(zkRealm)) {
```

Comment:
concurrentMap, we should use  PutIfAbsent(), no need to reinvent the wheel, which may not be right.

See here. https://stackoverflow.com/questions/10743622/concurrenthashmap-avoid-extra-object-creation-with-putifabsent

Reply:
@kaisun2000 Thanks for the suggestion. I am glad we got to this point.
- Did you mean `putIfAbsent(zkRealm, createZkClient(zkRealm))`? I did thought about it. However, it actually does not satisfy my expectation. `putIfAbsent(zkRealm, createZkClient(zkRealm))` runs `createZkClient(zkRealm)` every time, no matter if zkRealm presents or not. As you already know, creating a zkclient costs a lot of overhead: creating native zk instance, zk event thread, wait until connected, etc.. We definitely don't want to create a zkclient every time even one is already available, right?
- Then is there another API that could solve the pain we have above so no new zkclient is created if there is already one? Yes, that is `computeIfAbsent()`. We can make one time init and get() together as clean as `map.computeIfAbsent(zkRealm, zkclient -> createZkClient())`. This looks pretty good, right? But my only concern is that the get performance of `computeIfAbsent()` is not as good as simply a `get(zkRealm)`. Because internally in ConcurrentHashMap's implementation, `computeIfAbsent()` always synchronized the node/table, while `get()` directly reads the `transient volatile table` **without** synchonized/lock, which has a better performance over `computeIfAbsent()`.
- 'get()' + `synchronized`. Back to our use case, read is much much more than create/write because creation is only one time for each ZkClient. And I think the synchronized code block here is correct. This way makes sure creation/put is thread-safe, one time initialization does not hurt too much performance because read much heavier than write in our case. 

So I believe `get()` + `synchronized` satisfy us better. What do you think?

Reply:
Good point. ComputeifAbsent() is better. This guarantees no leaking of DedicatedZkClient. Note, client created must be close(). PutIfAbsent has a small chance of leaking though.

Get() + synchronized has its pitfall too. This is double-locking pattern. If test a simple value null or not, you need to make the value volatile. But for this one, I am not sure. 

So my suggestion is to use ComputeIfAbsent, and if possible, add a test case to make sure we don't leak zkClient (garbage collected, but did not call close())

Reply:
As we discussed, we will keep current implementation, because:
- double-check lock here for the concurrent map is thread safe, as `computeIfAbsent()`
- `get()` is faster than `computeIfAbsent()` for getting an existing key, as we analyzed the code and did performance experiments to prove it
- synchronized `_zkRealmToZkClientMap` also protects the map from adding new zkclient, when `close()` is called.

Reply:
This good implementation.  Please add a comment here why this is safe and performant ( https://stackoverflow.com/questions/7003239/double-checked-locking-with-concurrentmap)